File: 15. 3Sum/main.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        result = []
        for n in range(len(nums) - 2):
            if n != 0 and nums[n] == nums[n-1]:
                continue

            i, j = n+1, len(nums) - 1
            target = -1 * nums[n]
            while i < j:
                if target == nums[i] + nums[j]:
                    result.append([nums[n], nums[i], nums[j]])
                    i += 1
                    j -= 1

                    while i < j and nums[i] == nums[i-1]:
                        i += 1
                    while j > i and nums[j] == nums[j+1]:
                        j -= 1
                        
                elif nums[i] + nums[j] > target:
                    j -= 1
                else:
                    i += 1
        
        return result

File: 15. 3Sum/main.go
func threeSum(nums []int) [][]int {
    res := make([][]int, 0)

	slices.Sort(nums)

	for i := 0; i < len(nums) - 2; i++ {
        if i != 0 && nums[i] == nums[i-1] {
            continue
        }
		
		target := -1 * nums[i]
		left, right := i + 1, len(nums) - 1
		for left < right {
			if nums[left] + nums[right] == target {
				res = append(res, []int{nums[i], nums[left], nums[right]})
				left++
				right--

                for left < right && nums[left] == nums[left-1] {
                    left++
                }

                for left < right && nums[right] == nums[right+1] {
                    right--
                }

			} else if nums[left] + nums[right] > target {
				right--
			} else {
				left++
			}
		}
	}

	return res
}